board/safety/safety_volkswagen.h

```diff
@@ -0,0 +1,169 @@
+
+  // Monitor Klemmen_Status_01.ZAS_Kl_15 for Terminal 15 (ignition-on) status, but we make no use of it at the moment.
+  if ((bus == 0) && (addr == MSG_KLEMMEN_STATUS_01)) {
+    vw_ignition_started = (GET_BYTE(to_push, 2) & 0x2) >> 1;
+  }
+
+  // Update driver input torque samples from EPS_01.Driver_Strain for absolute torque, and EPS_01.Driver_Strain_VZ
+  // for the direction.
+  if ((bus == 0) && (addr == MSG_EPS_01)) {
+    int torque_driver_new = GET_BYTE(to_push, 5) | ((GET_BYTE(to_push, 6) & 0x1F) << 8);
+    int sign = (GET_BYTE(to_push, 6) & 0x80) >> 7;
+    if (sign == 1) {
+      torque_driver_new *= -1;
+    }
+
+    update_sample(&vw_torque_driver, torque_driver_new);
+  }
+
+  // Monitor ACC_06.ACC_Status_ACC for stock ACC status. Because the current MQB port is lateral-only, OP's control
+  // allowed state is directly driven by stock ACC engagement. Permit the ACC message to come from either bus, in
+  // order to accommodate future camera-side integrations if needed.
+  if (addr == MSG_ACC_06) {
+    int acc_status = (GET_BYTE(to_push,7) & 0x70) >> 4;
+    controls_allowed = ((acc_status == 3) || (acc_status == 4) || (acc_status == 5)) ? 1 : 0;
+  }
+}
+
+static int volkswagen_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
+  int addr = GET_ADDR(to_send);
+  int tx = 1;
+
+  // Safety check for HCA_01 Heading Control Assist torque.
+  if (addr == MSG_HCA_01) {
+    bool violation = false;
+
+    int desired_torque = GET_BYTE(to_send, 2) | ((GET_BYTE(to_send, 3) & 0x3F) << 8);
+    int sign = (GET_BYTE(to_send, 3) & 0x80) >> 7;
+    if (sign == 1) {
+      desired_torque *= -1;
+    }
+
+    uint32_t ts = TIM2->CNT;
+
+    if (controls_allowed) {
+
+      // *** global torque limit check ***
+      violation |= max_limit_check(desired_torque, VW_MAX_STEER, -VW_MAX_STEER);
+
+      // *** torque rate limit check ***
+      violation |= driver_limit_check(desired_torque, vw_desired_torque_last, &vw_torque_driver,
+        VW_MAX_STEER, VW_MAX_RATE_UP, VW_MAX_RATE_DOWN,
+        VW_DRIVER_TORQUE_ALLOWANCE, VW_DRIVER_TORQUE_FACTOR);
+      vw_desired_torque_last = desired_torque;
+
+      // *** torque real time rate limit check ***
+      violation |= rt_rate_limit_check(desired_torque, vw_rt_torque_last, VW_MAX_RT_DELTA);
+
+      // every RT_INTERVAL set the new limits
+      uint32_t ts_elapsed = get_ts_elapsed(ts, vw_ts_last);
+      if (ts_elapsed > VW_RT_INTERVAL) {
+        vw_rt_torque_last = desired_torque;
+        vw_ts_last = ts;
+      }
+    }
+
+    // no torque if controls is not allowed
+    if (!controls_allowed && (desired_torque != 0)) {
+      violation = true;
+    }
+
+    // reset to 0 if either controls is not allowed or there's a violation
+    if (violation || !controls_allowed) {
+      vw_desired_torque_last = 0;
+      vw_rt_torque_last = 0;
+      vw_ts_last = ts;
+    }
+
+    if (violation) {
+      tx = 0;
+    }
+
+  }
+
+  // 1 allows the message through
+  return tx;
```

Comment:
The violation->tx thing is a little redundant/overkill in our case, because our only no-transmit condition is a torque violation at this time. But, "it's the way everyone else does it" for convenience of review and audit, and we're likely to grow other conditions later.

Reply:
yeah, I prefer keeping the same format, even if it's a little redundant